api/v1alpha1/gateway_types.go

```diff
@@ -179,56 +182,101 @@ type ListenerTLS struct {
 
 // GatewayStatus defines the observed state of Gateway
 type GatewayStatus struct {
-	// TODO overall status
-
-	// Listeners is the status for each listener block in the
-	// Spec. The status for a given block will match the order as
-	// declared in the Spec, e.g. the status for Spec.Listeners[3]
-	// will be in Status.Listeners[3].
+	// Conditions describe the current conditions of the Gateway.
+	Conditions []GatewayCondition `json:"conditions"`
+	// Listeners provide status for each listener defined in the Spec. The name
+	// in ListenerStatus refers to the corresponding Listener of the same name.
 	Listeners []ListenerStatus `json:"listeners"`
-	// Routes is the status for each attached route to the
-	// Gateway. The status for a given route will match the orer as
-	// declared in the Spec, e.g. the status for Spec.Routes[3] will
-	// be in Status.Routes[3].
-	Routes []GatewayRouteStatus `json:"routes"`
+}
+
+// GatewayConditionType is a type of condition associated with a Gateway.
+type GatewayConditionType string
+
+const (
+	// ConditionNoSuchGatewayClass indicates that the specified GatewayClass
+	// does not exist.
+	ConditionNoSuchGatewayClass GatewayConditionType = "NoSuchGatewayClass"
+	// ConditionGatewayNotScheduled indicates that the Gateway has not been
+	// scheduled.
+	ConditionGatewayNotScheduled GatewayConditionType = "GatewayNotScheduled"
+	// ConditionListenersNotReady indicates that at least one of the specified
+	// listeners is not ready. If this condition has a status of True, a more
+	// detailed ListenerCondition should be present in the corresponding
+	// ListenerStatus.
+	ConditionListenersNotReady GatewayConditionType = "ListenersNotReady"
+	// ConditionInvalidListeners indicates that at least one of the specified
+	// listeners is invalid. If this condition has a status of True, a more
+	// detailed ListenerCondition should be present in the corresponding
+	// ListenerStatus.
+	ConditionInvalidListeners GatewayConditionType = "InvalidListeners"
+	// ConditionRoutesNotReady indicates that at least one of the specified
+	// routes is not ready.
+	ConditionRoutesNotReady GatewayConditionType = "RoutesNotReady"
+	// ConditionInvalidRoutes indicates that at least one of the specified
+	// routes is invalid.
+	ConditionInvalidRoutes GatewayConditionType = "InvalidRoutes"
+)
+
+// GatewayCondition is an error status for a given route.
+type GatewayCondition struct {
+	// Type indicates the type of condition.
+	Type GatewayConditionType `json:"type"`
+	// Status describes the current state of this condition. Can be "True",
+	// "False", or "Unknown".
+	Status core.ConditionStatus `json:"status"`
+	// Message is a human-understandable message describing the condition.
+	Message string `json:"message"`
```

Comment:
It was, but that was only because I'd seen an example of conditions used that way. It seems like message is optional most of the time though. Updated to make it optional here as well.